### ISSUE TITLE
Matrix-free fieldsplits involving R blocks deadlock in parallel

### ISSUE THREAD

**Issue by pefarrell** [firedrakeproject/firedrake#1415]
Firedrake cannot unfortunately assemble linear systems involving R blocks (to use LU). I therefore resort to Schur complement preconditioners that split onto the R block, use LU for the top-left block, and apply GMRES without any preconditioner to the (1x1) Schur complement system.

This strategy works successfully in serial but deadlocks in parallel. Consider the following code to solve the Moore-Spence system identifying the first pitchfork bifurcation point of the Euler elastica:

```
from defcon import *
from firedrake import *
from firedrake.petsc import PETSc
from slepc4py import SLEPc


msh = IntervalMesh(1000, 1)
V = FunctionSpace(msh, "CG", 1)
R = FunctionSpace(msh, "R", 0)

#elastica residual
def residual(theta, lmbda, ttheta):
    return inner(grad(theta), grad(ttheta))*dx - lmbda**2*sin(theta)*ttheta*dx

th = Function(V)
x = SpatialCoordinate(msh)[0]
tth = TestFunction(V)
lm = Constant(3.142)


# Using guess for parameter lm, solve for state theta (th)
A = residual(th, lm, tth)
bcs = [DirichletBC(V, 0.0, "on_boundary")]
solve(A==0, th, bcs=bcs)

#Now solve eigenvalue problem for $F_u(u, \lambda)\phi = r\phi$
#Want eigenmode phi with minimal eigenvalue r
B = derivative(residual(th, lm, TestFunction(V)),th, TrialFunction(V))

petsc_M = assemble(inner(TestFunction(V), TrialFunction(V))*dx, bcs=bcs).M.handle
petsc_B = assemble(B, bcs=bcs).M.handle

num_eigenvalues = 1

opts = PETSc.Options()
opts.setValue("eps_target_magnitude", None)
opts.setValue("eps_target", 0)
opts.setValue("st_type", "sinvert")

es = SLEPc.EPS().create(comm=COMM_WORLD)
es.setDimensions(num_eigenvalues)
es.setOperators(petsc_B, petsc_M)
es.setProblemType(SLEPc.EPS.ProblemType.GHEP)
es.setFromOptions()
es.solve()

ev_re, ev_im = petsc_B.getVecs()
r = es.getEigenpair(0, ev_re, ev_im)
eigenmode = Function(V)
eigenmode.vector().set_local(ev_re)

Z = MixedFunctionSpace([V,R,V])

# Set initial guesses for state, parameter, null eigenmode
z = Function(Z)
z.split()[0].assign(th)
z.split()[1].assign(lm)
z.split()[2].assign(eigenmode)

# Write Moore-Spence system of equations
theta, lmbda, phi = split(z)
ttheta, tlmbda, tphi = TestFunctions(Z)
F1 = residual(theta, lmbda, ttheta)
F2 = derivative(residual(theta, lmbda, tphi), z, as_vector([phi, 0, 0]))
F3 = (inner(phi, phi) - 1)*tlmbda*dx

F = F1 + F2 + F3

bcs = [DirichletBC(Z.sub(0), 0.0, "on_boundary"), DirichletBC(Z.sub(2), 0.0, "on_boundary")]

# Need to fieldsplit onto the real variable as assembly doesn't work with R
sp = {
    "mat_type": "matfree",
    "snes_type": "newtonls",
    "snes_monitor": None,
    "snes_converged_reason": None,
    #"snes_atol": 1.0e-6,
    "snes_linesearch_type": "basic",
    "ksp_type": "fgmres",
    "ksp_monitor_true_residual": None,
    "ksp_max_it": 10,
    "pc_type": "fieldsplit",
    "pc_fieldsplit_type": "schur",
    "pc_fieldsplit_schur_fact_type": "full",
    "pc_fieldsplit_0_fields": "0,2",
    "pc_fieldsplit_1_fields": "1",
    "fieldsplit_0_ksp_type": "preonly",
    "fieldsplit_0_pc_type": "python",
    "fieldsplit_0_pc_python_type": "firedrake.AssembledPC",
    "fieldsplit_0_assembled_pc_type": "lu",
    "fieldsplit_0_assembled_pc_factor_mat_solver_type": "mumps",
    "fieldsplit_1_ksp_type": "gmres",
    "fieldsplit_1_ksp_monitor_true_residual": None,
    "fieldsplit_1_ksp_max_it": 1,
    "fieldsplit_1_ksp_convergence_test": "skip",
    "fieldsplit_1_pc_type": "none",
    }


solve(F==0, z, bcs=bcs, solver_parameters=sp)
with z.sub(1).dat.vec_ro as x:
    print("Got parameter: %s" % x.array[0])

```

In serial, this works:

```
[farrellp@colossus:/tmp]$ python bug.py 
  0 SNES Function norm 8.068486260225e-05 
    Residual norms for firedrake_1_ solve.
    0 KSP unpreconditioned resid norm 8.068486260225e-05 true resid norm 8.068486260225e-05 ||r(i)||/||b|| 1.000000000000e+00
      Residual norms for firedrake_1_fieldsplit_1_ solve.
      0 KSP preconditioned resid norm 2.478779360572e+04 true resid norm 2.478779360572e+04 ||r(i)||/||b|| 1.000000000000e+00
      1 KSP preconditioned resid norm -0.000000000000e+00 true resid norm 3.637978807092e-11 ||r(i)||/||b|| 1.467649305524e-15
    1 KSP unpreconditioned resid norm 6.817578228650e-15 true resid norm 6.817578228483e-15 ||r(i)||/||b|| 8.449637278422e-11
  1 SNES Function norm 5.213294228755e-09 
    Residual norms for firedrake_1_ solve.
    0 KSP unpreconditioned resid norm 5.213294228755e-09 true resid norm 5.213294228755e-09 ||r(i)||/||b|| 1.000000000000e+00
      Residual norms for firedrake_1_fieldsplit_1_ solve.
      0 KSP preconditioned resid norm 3.835551224635e+08 true resid norm 3.835551224635e+08 ||r(i)||/||b|| 1.000000000000e+00
      1 KSP preconditioned resid norm -0.000000000000e+00 true resid norm 4.768371582031e-07 ||r(i)||/||b|| 1.243203728164e-15
    1 KSP unpreconditioned resid norm 2.916691102913e-15 true resid norm 2.916691102913e-15 ||r(i)||/||b|| 5.594717993903e-07
  2 SNES Function norm 3.790835008619e-12 
Nonlinear firedrake_1_ solve converged due to CONVERGED_SNORM_RELATIVE iterations 2
Got parameter: 3.141593945518148
```

(the true value of the parameter is pi), but in parallel it deadlocks:
```

[farrellp@colossus:/tmp]$ mpiexec -n 2 python bug.py 
  0 SNES Function norm 8.068486260193e-05 
...
```

Investigating, the problem is the call to `find_sub_block(row_is, row_ises)` in `firedrake/matrix_free/operators.py`. The second time line 32 of that file is hit, on one process `lsize == 0` and `iset.getLocalSize() == 0`, so it goes into creating `tmp = PETSc.IS().createGeneral(indices[:lsize], comm=comm)` with a collective operation. However, on the other process, `lsize == 1` (the one that has the real dof) but `iset.getLocalSize() == 0`, so it hits the `continue` on line 33 and skips to the next index set, missing the collective call.

The following patch fixes the deadlock

```
[farrellp@colossus:~/local/firedrake/firedrake-dev-20190321-mpich/src/firedrake] (git:master) $ git diff
diff --git a/firedrake/matrix_free/operators.py b/firedrake/matrix_free/operators.py
index 84efbff7..e967010e 100644
--- a/firedrake/matrix_free/operators.py
+++ b/firedrake/matrix_free/operators.py
@@ -29,8 +29,6 @@ def find_sub_block(iset, ises):
             if i in sfound:
                 continue
             lsize = iset_.getLocalSize()
-            if lsize > iset.getLocalSize():
-                continue
             indices = iset.indices
             tmp = PETSc.IS().createGeneral(indices[:lsize], comm=comm)
             if tmp.equal(iset_):

```

but I don't understand the code well enough to assert this is definitely the right thing to do.

**Comment by pefarrell** [comment]
Speculative fix proposed in pull request #1420.

**Comment by pefarrell** [comment]
This is now fixed in master.
